Preserve expanded path when selecting new version

mirefly opened the pull request:
Fixes https://github.com/mozilla/addons-code-manager/issues/1248

This is an attempt to refactor `expandedPaths` out of `version` object in a progressive way according to https://github.com/mozilla/addons-code-manager/pull/1100#issuecomment-549474948

mirefly commented:
@kumar303 This patch is ready for a review.

kumar303 commented on src/reducers/versions.tsx:
```diff
@@ -195,13 +195,16 @@ type VersionAddon = {
   version: string;
+  // TODO: remove `expandedPaths`, `selectedPath` and `visibleSelectedPath`
+  // from `version` object once no code depends on these
+  // https://github.com/mozilla/addons-code-manager/pull/1204
```
I made https://github.com/mozilla/addons-code-manager/issues/1218 so we don't lose track of it:

```suggestion
  // See https://github.com/mozilla/addons-code-manager/issues/1218
```

mirefly commented on src/reducers/versions.tsx:
```diff
@@ -1310,17 +1322,24 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
-      const { expandedPaths } = version;
+      const { expandedPaths } = state;
+
+      if (expandedPaths === undefined) {
```
I considered this before, but finally choose to use `undefined`.

The initial `expandedPaths` needs to be a special one, so that we can know if the old `expandedPaths` should be preserved or not when load a new version.

Using empty array as initial value works in this case, since `expandedPaths` always have at least one element `~ROOT~` after initialized. However, it looks bad to have such implicitly dependence on the fact that we have `~ROOT~` which cannot moved out of `expandedPaths`. And It is not that readable to use `expandedPaths.length() === 0` to determine if it is initialized.

> Would that work for the other states?

Do you mean `selectedPath` and `visibleSelectedPath`? I guess they need to be defined as `string|undefined`.

kumar303 replied:
> The initial `expandedPaths` needs to be a special one, so that we can know if the old `expandedPaths` should be preserved or not when load a new version.

I don't see anything resetting `expandedPaths` to `undefined` after loading a version so does switching versions work as it should? After https://github.com/mozilla/addons-code-manager/pull/1204#discussion_r343889489 gets addressed then I can look at that scenario in more detail.

> > Would that work for the other states?
> 
> Do you mean `selectedPath` and `visibleSelectedPath`? I guess they need to be defined as `string|undefined`.

I was just asking if using `string[]` for `expandedPaths` would work for the other features we need to support.

mirefly replied:
> I was just asking if using string[] for expandedPaths would work for the other features we need to support.

Yes. We only have `expandedPaths === undefined` before loading the first version (we do not reset it to `undefined`), so using [] as initial value should not break other features.

kumar303 commented:
Let me know when it's ready for another look.

mirefly commented on src/reducers/versions.tsx:
```diff
@@ -1157,11 +1162,15 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       const { version } = action.payload;
 
+      const expandedPaths =
+        state.expandedPaths || getParentFolders(version.file.selected_file);
```
I added the filtering out here, and some tests for this. I realized it may be a bit more tricky than I thought... When switching to a version which has been loaded before, `loadVersionInfo` will not be dispatched, so the paths that do not exist in new version cannot be removed unless we introduce a new action to do this. On the other hand, it looks not harmful if we just leave those nonexistent paths there (or I missed something?).

kumar303 replied:
Would it help to fix the paths when `setCurrentVersionId` gets dispatched?

https://github.com/mozilla/addons-code-manager/blob/3afafcd01ce6051f7cff6bc1dbee9f02ab021828/src/reducers/versions.tsx#L1459-L1465

I'm not sure how well `setCurrentVersionId` is supported in places like the compare page, tbh.

You could be right, maybe it doesn't matter if there are foreign paths in `expandedPaths`. My one concern is that switching between lots of versions will start to introduce a memory leak. However, a fix for that could be [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

mirefly replied:
> Would it help to fix the paths when setCurrentVersionId gets dispatched?

Great. It looks like a good place to fix `expandedPath`. I come up with the following patch, which moves path filtering into `setCurrentVersionId`. It is a bit strange to let `setCurrentVersionId` do such clean-up, but it works. Do you want me to proceed in this direction?
```diff
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index a437f60..abdbad0 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1164,9 +1164,8 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       const { expandedPaths: paths } = state;
       const entries = Object.keys(version.file.entries);
 
-      const expandedPaths = paths
-        ? paths.filter((e) => entries.includes(e))
-        : getParentFolders(version.file.selected_file);
+      const expandedPaths =
+        paths || getParentFolders(version.file.selected_file);
 
       return {
         ...state,
@@ -1483,6 +1482,26 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
     }
     case getType(actions.setCurrentVersionId): {
       const { versionId } = action.payload;
+      let { expandedPaths } = state;
+      const version = getVersionInfo(state, versionId);
+
+      if (version && expandedPaths) {
+        const paths = version.entries.map((e) => e.path);
+        expandedPaths = expandedPaths.filter((e) => paths.includes(e));
+
+        return {
+          ...state,
+          currentVersionId: versionId,
+          expandedPaths,
+          versionInfo: {
+            [versionId]: {
+              ...version,
+              expandedPaths,
+            },
+          },
+        };
+      }
+
       return {
         ...state,
         currentVersionId: versionId,

```

kumar303 replied:
Yeah, I think it's worth proceeding. I double checked and it looks like `setCurrentVersionId()` is in fact dispatched in all the right places. I wouldn't know for sure until trying things out, though.

kumar303 replied:
I wouldn't worry about supporting `versionInfo` in this new reducer because we don't need to invest in new features for the `version.expandedPaths` property.

mirefly commented:
@kumar303 It is ready for another look, even if https://github.com/mozilla/addons-code-manager/pull/1204#discussion_r343889489 is not completely addressed...

mirefly commented:
It is now updated to clean up `expandedPaths` in `setCurrentVersionId`.

kumar303 commented on src/reducers/versions.tsx:
```diff
@@ -1458,6 +1481,20 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
+      const version = getVersionInfo(state, versionId);
+
+      if (version && expandedPaths) {
+        const paths = version.entries.map((e) => e.path);
```
It's not a big deal but you could wrap this array in a [set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) since it's used only for lookups.

kumar303 commented on src/reducers/versions.tsx:
```diff
@@ -1458,6 +1481,20 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
+
+      if (version && expandedPaths) {
+        const paths = version.entries.map((e) => e.path);
+        expandedPaths = expandedPaths.filter((e) => paths.includes(e));
```
This looks right but the `loadVersionInfo` reducer also needs to access this filter. I suggest moving it to a shared function and making unit tests for it as a function.

As an example, `fetchDiff` relies on [`loadVersionInfo`](https://github.com/mozilla/addons-code-manager/blob/9b82a1070efb275eef65a3975fc9732cc27f2100/src/reducers/versions.tsx#L1001) to switch versions and filtering expanded paths would be important at that point. The same thunk dispatches `setCurrentVersionId` earlier on but that happens before the new `version` has loaded.

mirefly replied:
FWIW, it looks that `setCurrentVersionId` is always (?) dispatched again after loading a new version when `_fetchDiff` for preloading next diff. I am fine with adding the filter in `loadVersionInfo` (maybe making it not depend on preloading), but it does not seem to have real benefit except filtering the paths once more.. How do you think about it?

kumar303 replied:
We can't depend on preloading since the user can click on any file path out of order. I think it would be safer to introduce a shared function for both reducers to use and it's ok if the paths get filtered twice. You won't have to duplicate any tests for this. I suggest testing it only in the shared function.

mirefly commented:
@kumar303 Thanks a lot for helping on this. I addressed all comments.
It looks like new released prettier 1.19 have some changes on formatting and make CI fail. Running `yarn prettier` changes many irrelevant files...

kumar303 commented:
@mirefly did you try rebasing on / merging with master? The prettier changes just landed in https://github.com/mozilla/addons-code-manager/pull/1225/files

kumar303 commented on src/reducers/versions.spec.tsx:
```diff
@@ -3135,4 +3240,33 @@ describe(__filename, () => {
+
+      expect(
+        getPrunedExpandedPaths(expandedPaths, createInternalVersion(version)),
+      ).toEqual([path1]);
```
I like how you made separate tests for inclusion / exclusion. As it is, this assertion covers both inclusion and exclusion so I suggest making this one just about inclusion. Example `expect(...).arrayContaining([path1])`

kumar303 commented on src/reducers/versions.spec.tsx:
```diff
@@ -3135,4 +3240,33 @@ describe(__filename, () => {
+
+      expect(
+        getPrunedExpandedPaths(expandedPaths, createInternalVersion(version)),
+      ).toEqual([path1]);
```
Similarly, this should just cover exclusion: `expect(...).not.arrayContaining([path2])`

kumar303 commented on src/reducers/versions.spec.tsx:
```diff
@@ -117,12 +119,84 @@ describe(__filename, () => {
       });
     });
 
+    it('prunes expandedPaths when load version info', () => {
```
This test duplicates the logic of `getPrunedExpandedPaths()` which isn't necessary because you've passed on the responsibility of that logic to the function. Adding duplicated tests slows down the test suite over time and also adds a maintenance burden because any change in `getPrunedExpandedPaths()` could cause you to update unrelated tests.

Instead, I'd like to see this test inject a stub, like this:

```diff
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 616f66d..4ef6128 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1154,6 +1154,7 @@ export const goToRelativeDiff = ({
 const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
   state = initialState,
   action,
+  { _getPrunedExpandedPaths = getPrunedExpandedPaths } = {},
 ): VersionsState => {
   switch (action.type) {
     case getType(actions.beginFetchVersion): {
@@ -1173,7 +1174,7 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       const { expandedPaths: expandedPathsInState } = state;
 
       const expandedPaths = expandedPathsInState
-        ? getPrunedExpandedPaths(expandedPathsInState, internalVersion)
+        ? _getPrunedExpandedPaths(expandedPathsInState, internalVersion)
         : getParentFolders(version.file.selected_file);
 
       return {
@@ -1498,7 +1499,7 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       const version = getVersionInfo(state, versionId);
 
       if (version && expandedPaths) {
-        expandedPaths = getPrunedExpandedPaths(expandedPaths, version);
+        expandedPaths = _getPrunedExpandedPaths(expandedPaths, version);
       }
 
       return {
```

The stub can be a spy that returns the *real* `getPrunedExpandedPaths()`. This will make it so you don't have to do any special set up to simulate its return value. Example:

```js
reducer(
  undefined,
  actions.loadVersionInfo({ version: oldVersion }),
  { _getPrunedExpandedPaths: jest.fn(getPrunedExpandedPaths) }
)
```

The test can then verify that both `expandedPaths` and `internalVersion` are passed to `getPrunedExpandedPaths`. That's all it needs to check because this passes all other responsibility on to `getPrunedExpandedPaths()`, which has its own tests.

kumar303 commented on src/reducers/versions.spec.tsx:
```diff
@@ -117,12 +119,84 @@ describe(__filename, () => {
+      );
+    });
+
+    it('prunes expanded paths when setCurrentVersionId is dispatched', () => {
```
This should use dependency injection as described above, to ensure that responsibility is passed on to `getPrunedExpandedPaths` correctly.

kumar303 commented on src/reducers/versions.tsx:
```diff
@@ -463,6 +468,14 @@ export const getVersionInfo = (versions: VersionsState, versionId: number) => {
+  version: Version,
+) => {
+  const paths = new Set(version.entries.map((e) => e.path));
+  return expandedPaths.filter((e) => paths.has(e));
```
Since `e` is not an entry (as it is above), I suggest picking a different variable like `p` (for path) or something.

mirefly commented:
> 
> @mirefly did you try rebasing on / merging with master? 

I forgot running `yarn install`... Now it is fine.

mirefly commented on src/reducers/versions.tsx:
```diff
@@ -1256,21 +1276,27 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
+      const { expandedPaths } = state;
+
+      if (!expandedPaths) {
+        throw new Error('Expanded paths are undefined');
```
https://github.com/mozilla/addons-code-manager/blob/d83d73b383bfdaac4c9f239f22ed67266fe27535/src/reducers/versions.tsx#L1275

We should not get this error in real, since we should get "missing version" error before that. To test this line, I have not way to prepare the state with a version loaded and expandedPaths undefined. Or should I manually set the expandedPaths `undefined`?

kumar303 replied:
Ah, right, that's tricky. It's ok to ignore it for now (TS covers it well). You can add this so that it's not distracting in the coverage report:

```diff
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 616f66d..58f22bf 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1279,6 +1279,10 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       const { expandedPaths } = state;
 
       if (!expandedPaths) {
+        // TODO: this is not possible to test until expanded paths are
+        // decoupled from versions.
+        // See https://github.com/mozilla/addons-code-manager/issues/1218
+        /* istanbul ignore next */
         throw new Error('Expanded paths are undefined');
       }
 
```

mirefly commented on src/reducers/versions.tsx:
```diff
@@ -1138,9 +1150,10 @@ export const goToRelativeDiff = ({
+  state: VersionsState | undefined = initialState,
+  action: ActionType<typeof actions>,
+  { _getPrunedExpandedPaths = getPrunedExpandedPaths } = {},
 ): VersionsState => {
```
I did not figure out how to reuse the generic type `Reducer<S,A>` here, hoping it is not a problem...

kumar303 commented on src/reducers/versions.spec.tsx:
```diff
@@ -117,12 +119,81 @@ describe(__filename, () => {
+      const path2 = 'path/2';
+      const expandedPaths = [path1, path2, ROOT_PATH];
+      const oldVersionId = nextUniqueId();
+      const oldVersion = createExternalVersionWithEntries(
```
This test sets up a situation where the expanded paths of an old version are filtered out according to a new version. However, that filtering logic is already tested [here](https://github.com/mozilla/addons-code-manager/blob/4efe46a9711b8491a5d2c4faf56720a7c41814ff/src/reducers/versions.spec.tsx#L3259) since it's part of the implementation of `getPrunedExpandedPaths()`. The reducer will be relying on that logic but all this test needs to do is assert that reducing the `loadVersionInfo` action calls `getPrunedExpandedPaths()` with:

1. the `expandedPaths` that exist in state; and
2. the newly created internal version object.

That's it. `getPrunedExpandedPath()` guarantees that it will prune paths given the correct input.

With that in mind, see if you can simplify this test setup. Think about how you only need to assert the two points I outlined above and not any of the logic contained *within* `getPrunedExpandedPath()`.

If you get stuck, let me know.

kumar303 commented on src/reducers/versions.spec.tsx:
```diff
@@ -117,12 +119,81 @@ describe(__filename, () => {
+      );
+    });
+
+    it('prunes expanded paths when setCurrentVersionId is dispatched', () => {
```
My same comment above applies to this test as well.

kumar303 commented on src/reducers/versions.tsx:
```diff
@@ -1138,9 +1150,10 @@ export const goToRelativeDiff = ({
   };
 };
 
-const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
```
Ah, whoops. I forgot that you need to define it as a function that returns a reducer. I can use the *hide whitespace* feature in github to review the patch since it will indent a lot of code.

<details>
<summary>Example patch</summary>

```diff
diff --git a/src/reducers/versions.spec.tsx b/src/reducers/versions.spec.tsx
index 2bc92d9..a110700 100644
--- a/src/reducers/versions.spec.tsx
+++ b/src/reducers/versions.spec.tsx
@@ -19,6 +19,7 @@ import reducer, {
   createInternalVersionAddon,
   createInternalVersionEntry,
   createInternalVersionFile,
+  createReducer,
   createVersionsMap,
   fetchDiff,
   fetchVersion,
@@ -143,12 +144,16 @@ describe(__filename, () => {
         { id: nextUniqueId() },
       );
 
-      let state = reducer(
+      const _reducer = createReducer({
+        _getPrunedExpandedPaths: jest.fn(getPrunedExpandedPaths),
+      })
+
+      let state = _reducer(
         undefined,
         actions.loadVersionInfo({ version: oldVersion }),
       );
-      state = reducer(state, actions.expandTree({ versionId: oldVersionId }));
-      state = reducer(state, actions.loadVersionInfo({ version: newVersion }));
+      state = _reducer(state, actions.expandTree({ versionId: oldVersionId }));
+      state = _reducer(state, actions.loadVersionInfo({ version: newVersion }));
 
       expect(state).toHaveProperty(
         'expandedPaths',
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 616f66d..94c5233 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1151,371 +1151,379 @@ export const goToRelativeDiff = ({
   };
 };
 
-const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
-  state = initialState,
-  action,
-): VersionsState => {
-  switch (action.type) {
-    case getType(actions.beginFetchVersion): {
-      const { versionId } = action.payload;
-
-      return {
-        ...state,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: undefined,
-        },
-      };
-    }
-    case getType(actions.loadVersionInfo): {
-      const { version } = action.payload;
-      const internalVersion = createInternalVersion(version);
-      const { expandedPaths: expandedPathsInState } = state;
-
-      const expandedPaths = expandedPathsInState
-        ? getPrunedExpandedPaths(expandedPathsInState, internalVersion)
-        : getParentFolders(version.file.selected_file);
-
-      return {
-        ...state,
-        expandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [version.id]: { ...internalVersion, expandedPaths },
-        },
-      };
-    }
-    case getType(actions.abortFetchVersion): {
-      const { versionId } = action.payload;
-
-      return {
-        ...state,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: null,
-        },
-      };
-    }
-    case getType(actions.loadEntryStatusMap): {
-      const { version, comparedToVersionId } = action.payload;
-      const key = getEntryStatusMapKey({
-        versionId: version.id,
-        comparedToVersionId,
-      });
-
-      return {
-        ...state,
-        entryStatusMaps: {
-          ...state.entryStatusMaps,
-          [key]: createEntryStatusMap(version),
-        },
-      };
-    }
-    case getType(actions.loadVersionFile): {
-      const { path, version } = action.payload;
-
-      return {
-        ...state,
-        versionFiles: {
-          ...state.versionFiles,
-          [version.id]: {
-            ...state.versionFiles[version.id],
-            [path]: createInternalVersionFile(version.file),
+export const createReducer = ({
+  _getPrunedExpandedPaths = getPrunedExpandedPaths,
+} = {}) => {
+  const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
+    state = initialState,
+    action,
+  ): VersionsState => {
+    switch (action.type) {
+      case getType(actions.beginFetchVersion): {
+        const { versionId } = action.payload;
+
+        return {
+          ...state,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: undefined,
           },
-        },
-        versionFilesLoading: {
-          ...state.versionFilesLoading,
-          [version.id]: {
-            ...state.versionFilesLoading[version.id],
-            [path]: false,
+        };
+      }
+      case getType(actions.loadVersionInfo): {
+        const { version } = action.payload;
+        const internalVersion = createInternalVersion(version);
+        const { expandedPaths: expandedPathsInState } = state;
+
+        const expandedPaths = expandedPathsInState
+          ? _getPrunedExpandedPaths(expandedPathsInState, internalVersion)
+          : getParentFolders(version.file.selected_file);
+
+        return {
+          ...state,
+          expandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [version.id]: { ...internalVersion, expandedPaths },
           },
-        },
-      };
-    }
-    case getType(actions.beginFetchVersionFile): {
-      const { path, versionId } = action.payload;
-
-      return {
-        ...state,
-        versionFilesLoading: {
-          ...state.versionFilesLoading,
-          [versionId]: {
-            ...state.versionFilesLoading[versionId],
-            [path]: true,
+        };
+      }
+      case getType(actions.abortFetchVersion): {
+        const { versionId } = action.payload;
+
+        return {
+          ...state,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: null,
           },
-        },
-      };
-    }
-    case getType(actions.abortFetchVersionFile): {
-      const { path, versionId } = action.payload;
-
-      return {
-        ...state,
-        versionFiles: {
-          ...state.versionFiles,
-          [versionId]: {
-            ...state.versionFiles[versionId],
-            [path]: null,
+        };
+      }
+      case getType(actions.loadEntryStatusMap): {
+        const { version, comparedToVersionId } = action.payload;
+        const key = getEntryStatusMapKey({
+          versionId: version.id,
+          comparedToVersionId,
+        });
+
+        return {
+          ...state,
+          entryStatusMaps: {
+            ...state.entryStatusMaps,
+            [key]: createEntryStatusMap(version),
           },
-        },
-        versionFilesLoading: {
-          ...state.versionFilesLoading,
-          [versionId]: {
-            ...state.versionFilesLoading[versionId],
-            [path]: false,
+        };
+      }
+      case getType(actions.loadVersionFile): {
+        const { path, version } = action.payload;
+
+        return {
+          ...state,
+          versionFiles: {
+            ...state.versionFiles,
+            [version.id]: {
+              ...state.versionFiles[version.id],
+              [path]: createInternalVersionFile(version.file),
+            },
           },
-        },
-      };
-    }
-    case getType(actions.updateSelectedPath): {
-      const { selectedPath, versionId } = action.payload;
-
-      const version = state.versionInfo[versionId];
-
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
+          versionFilesLoading: {
+            ...state.versionFilesLoading,
+            [version.id]: {
+              ...state.versionFilesLoading[version.id],
+              [path]: false,
+            },
+          },
+        };
       }
-
-      const { expandedPaths } = state;
-
-      if (!expandedPaths) {
-        throw new Error('Expanded paths are undefined');
+      case getType(actions.beginFetchVersionFile): {
+        const { path, versionId } = action.payload;
+
+        return {
+          ...state,
+          versionFilesLoading: {
+            ...state.versionFilesLoading,
+            [versionId]: {
+              ...state.versionFilesLoading[versionId],
+              [path]: true,
+            },
+          },
+        };
       }
-
-      const parents = getParentFolders(selectedPath);
-      const newExpandedPaths = [
-        ...expandedPaths,
-        ...parents.filter((newPath) => !expandedPaths.includes(newPath)),
-      ];
-
-      return {
-        ...state,
-        expandedPaths: newExpandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths: newExpandedPaths,
-            selectedPath,
+      case getType(actions.abortFetchVersionFile): {
+        const { path, versionId } = action.payload;
+
+        return {
+          ...state,
+          versionFiles: {
+            ...state.versionFiles,
+            [versionId]: {
+              ...state.versionFiles[versionId],
+              [path]: null,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.setVisibleSelectedPath): {
-      const { path, versionId } = action.payload;
+          versionFilesLoading: {
+            ...state.versionFilesLoading,
+            [versionId]: {
+              ...state.versionFilesLoading[versionId],
+              [path]: false,
+            },
+          },
+        };
+      }
+      case getType(actions.updateSelectedPath): {
+        const { selectedPath, versionId } = action.payload;
 
-      const version = state.versionInfo[versionId];
+        const version = state.versionInfo[versionId];
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
-      }
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      if (path && !version.entries.find((e) => e.path === path)) {
-        throw new Error(
-          `Path "${path}" is an unknown path for version ID ${versionId}`,
-        );
-      }
+        const { expandedPaths } = state;
 
-      return {
-        ...state,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            visibleSelectedPath: path,
+        if (!expandedPaths) {
+          throw new Error('Expanded paths are undefined');
+        }
+
+        const parents = getParentFolders(selectedPath);
+        const newExpandedPaths = [
+          ...expandedPaths,
+          ...parents.filter((newPath) => !expandedPaths.includes(newPath)),
+        ];
+
+        return {
+          ...state,
+          expandedPaths: newExpandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths: newExpandedPaths,
+              selectedPath,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.toggleExpandedPath): {
-      const { path, versionId } = action.payload;
+        };
+      }
+      case getType(actions.setVisibleSelectedPath): {
+        const { path, versionId } = action.payload;
 
-      const version = state.versionInfo[versionId];
+        const version = state.versionInfo[versionId];
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
-      }
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      const { expandedPaths } = state;
+        if (path && !version.entries.find((e) => e.path === path)) {
+          throw new Error(
+            `Path "${path}" is an unknown path for version ID ${versionId}`,
+          );
+        }
 
-      if (expandedPaths === undefined) {
-        throw new Error('ExpandedPaths is undefined');
+        return {
+          ...state,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              visibleSelectedPath: path,
+            },
+          },
+        };
       }
+      case getType(actions.toggleExpandedPath): {
+        const { path, versionId } = action.payload;
 
-      const newExpandedPaths = expandedPaths.includes(path)
-        ? expandedPaths.filter((storedPath) => path !== storedPath)
-        : [...expandedPaths, path];
+        const version = state.versionInfo[versionId];
 
-      return {
-        ...state,
-        expandedPaths: newExpandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths: newExpandedPaths,
-          },
-        },
-      };
-    }
-    case getType(actions.expandTree): {
-      const { versionId } = action.payload;
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      const version = state.versionInfo[versionId];
+        const { expandedPaths } = state;
+
+        if (expandedPaths === undefined) {
+          throw new Error('ExpandedPaths is undefined');
+        }
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
+        const newExpandedPaths = expandedPaths.includes(path)
+          ? expandedPaths.filter((storedPath) => path !== storedPath)
+          : [...expandedPaths, path];
+
+        return {
+          ...state,
+          expandedPaths: newExpandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths: newExpandedPaths,
+            },
+          },
+        };
       }
+      case getType(actions.expandTree): {
+        const { versionId } = action.payload;
 
-      const expandedPaths = version.entries
-        .filter((entry) => entry.type === 'directory')
-        .map((entry) => entry.path);
-      expandedPaths.push(ROOT_PATH);
+        const version = state.versionInfo[versionId];
 
-      return {
-        ...state,
-        expandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths,
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
+
+        const expandedPaths = version.entries
+          .filter((entry) => entry.type === 'directory')
+          .map((entry) => entry.path);
+        expandedPaths.push(ROOT_PATH);
+
+        return {
+          ...state,
+          expandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.collapseTree): {
-      const { versionId } = action.payload;
+        };
+      }
+      case getType(actions.collapseTree): {
+        const { versionId } = action.payload;
 
-      const version = state.versionInfo[versionId];
-      const expandedPaths = [ROOT_PATH];
+        const version = state.versionInfo[versionId];
+        const expandedPaths = [ROOT_PATH];
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
-      }
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      return {
-        ...state,
-        expandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths,
+        return {
+          ...state,
+          expandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.loadVersionsList): {
-      const { addonId, versions } = action.payload;
+        };
+      }
+      case getType(actions.loadVersionsList): {
+        const { addonId, versions } = action.payload;
+
+        return {
+          ...state,
+          byAddonId: {
+            ...state.byAddonId,
+            [addonId]: createVersionsMap(versions),
+          },
+        };
+      }
+      case getType(actions.beginFetchDiff): {
+        const key = getCompareInfoKey(action.payload);
+
+        return {
+          ...state,
+          compareInfo: {
+            ...state.compareInfo,
+            [key]: undefined,
+          },
+          compareInfoIsLoading: {
+            ...state.compareInfoIsLoading,
+            [key]: true,
+          },
+        };
+      }
+      case getType(actions.abortFetchDiff): {
+        const key = getCompareInfoKey(action.payload);
+
+        return {
+          ...state,
+          compareInfo: {
+            ...state.compareInfo,
+            [key]: null,
+          },
+          compareInfoIsLoading: {
+            ...state.compareInfoIsLoading,
+            [key]: false,
+          },
+        };
+      }
+      case getType(actions.loadDiff): {
+        const {
+          addonId,
+          baseVersionId,
+          headVersionId,
+          path,
+          version,
+        } = action.payload;
 
-      return {
-        ...state,
-        byAddonId: {
-          ...state.byAddonId,
-          [addonId]: createVersionsMap(versions),
-        },
-      };
-    }
-    case getType(actions.beginFetchDiff): {
-      const key = getCompareInfoKey(action.payload);
+        const compareInfoKey = getCompareInfoKey({
+          addonId,
+          baseVersionId,
+          headVersionId,
+          path,
+        });
 
-      return {
-        ...state,
-        compareInfo: {
-          ...state.compareInfo,
-          [key]: undefined,
-        },
-        compareInfoIsLoading: {
-          ...state.compareInfoIsLoading,
-          [key]: true,
-        },
-      };
-    }
-    case getType(actions.abortFetchDiff): {
-      const key = getCompareInfoKey(action.payload);
+        const headVersion = getVersionInfo(state, headVersionId);
+        if (!headVersion) {
+          throw new Error(
+            `Version missing for headVersionId: ${headVersionId}`,
+          );
+        }
 
-      return {
-        ...state,
-        compareInfo: {
-          ...state.compareInfo,
-          [key]: null,
-        },
-        compareInfoIsLoading: {
-          ...state.compareInfoIsLoading,
-          [key]: false,
-        },
-      };
-    }
-    case getType(actions.loadDiff): {
-      const {
-        addonId,
-        baseVersionId,
-        headVersionId,
-        path,
-        version,
-      } = action.payload;
+        const { entries, selectedPath } = headVersion;
+        const entry = entries.find((e) => e.path === selectedPath);
 
-      const compareInfoKey = getCompareInfoKey({
-        addonId,
-        baseVersionId,
-        headVersionId,
-        path,
-      });
+        if (!entry) {
+          throw new Error(`Entry missing for headVersionId: ${headVersionId}`);
+        }
 
-      const headVersion = getVersionInfo(state, headVersionId);
-      if (!headVersion) {
-        throw new Error(`Version missing for headVersionId: ${headVersionId}`);
+        return {
+          ...state,
+          compareInfo: {
+            ...state.compareInfo,
+            [compareInfoKey]: createInternalCompareInfo({
+              baseVersionId,
+              headVersionId,
+              entry,
+              version,
+            }),
+          },
+          compareInfoIsLoading: {
+            ...state.compareInfoIsLoading,
+            [compareInfoKey]: false,
+          },
+        };
       }
+      case getType(actions.setCurrentVersionId): {
+        const { versionId } = action.payload;
+        let { expandedPaths } = state;
+        const version = getVersionInfo(state, versionId);
 
-      const { entries, selectedPath } = headVersion;
-      const entry = entries.find((e) => e.path === selectedPath);
+        if (version && expandedPaths) {
+          expandedPaths = _getPrunedExpandedPaths(expandedPaths, version);
+        }
 
-      if (!entry) {
-        throw new Error(`Entry missing for headVersionId: ${headVersionId}`);
+        return {
+          ...state,
+          currentVersionId: versionId,
+          expandedPaths,
+        };
       }
-
-      return {
-        ...state,
-        compareInfo: {
-          ...state.compareInfo,
-          [compareInfoKey]: createInternalCompareInfo({
-            baseVersionId,
-            headVersionId,
-            entry,
-            version,
-          }),
-        },
-        compareInfoIsLoading: {
-          ...state.compareInfoIsLoading,
-          [compareInfoKey]: false,
-        },
-      };
-    }
-    case getType(actions.setCurrentVersionId): {
-      const { versionId } = action.payload;
-      let { expandedPaths } = state;
-      const version = getVersionInfo(state, versionId);
-
-      if (version && expandedPaths) {
-        expandedPaths = getPrunedExpandedPaths(expandedPaths, version);
+      case getType(actions.unsetCurrentVersionId): {
+        return {
+          ...state,
+          currentVersionId: false,
+        };
       }
-
-      return {
-        ...state,
-        currentVersionId: versionId,
-        expandedPaths,
-      };
-    }
-    case getType(actions.unsetCurrentVersionId): {
-      return {
-        ...state,
-        currentVersionId: false,
-      };
+      default:
+        return state;
     }
-    default:
-      return state;
-  }
+  };
+
+  return reducer;
 };
 
-export default reducer;
+export default createReducer();
```

</details>

mirefly commented on src/reducers/versions.spec.tsx:
```diff
@@ -117,12 +120,48 @@ describe(__filename, () => {
+
+      expect(_getPrunedExpandedPaths).toHaveBeenCalledWith(
+        state.expandedPaths,
+        expect.objectContaining({ id: newVersion.id }),
```
This line should be `createInternalVersion(newVersion)` after we remove `version.expandedPaths` completely #1218 .

kumar303 replied:
> This line should be `createInternalVersion(newVersion)` after we remove `version.expandedPaths` completely #1218 .

Actually, I think it's better as is because it does not depend on the implementation of `createInternalVersion()`.

mirefly commented on src/reducers/versions.spec.tsx:
```diff
@@ -117,12 +120,48 @@ describe(__filename, () => {
+      _reducer(state, actions.loadVersionInfo({ version: newVersion }));
+
+      expect(_getPrunedExpandedPaths).toHaveBeenCalledWith(
+        state.expandedPaths,
```
I am not very sure to have `state.expandedPaths` (which might be too simple like [] or [ROOT]) in this assertion, but if I predefine `expandedPaths = [path1, path2]`, then I will need to create a fake version with entries and dispatch `expandTree`.

kumar303 replied:
> I am not very sure to have `state.expandedPaths` ...

👍 

This is a good value to use for the assertion because the `expandedPaths` contents don't matter here. The assertion is just checking that the value gets passed to `getPrunedExpandedPaths()`.

mirefly commented:
Thanks @kumar303 . I updated the patch.
